File: rules/debugflag.js
const fs = require('fs');
const vscode = require('vscode');

var smell = {

    detect : (fileName, token) => {
        try{
            if(token.hasOwnProperty("line")) var lineno = token.line;
            if(token.hasOwnProperty("type")) var tokenType = token.type;
            if(token.hasOwnProperty("name")) var name = token.name;
            if(token.hasOwnProperty("value")) var value = token.value;
    
            const MSG = 'debug facility turned on'        
            const WARNING_MSG = MSG+' at line '+ lineno;
            const restrictedNames = ['debug','debug_propagate_exceptions','propagate_exceptions'];
    
            if(tokenType == "variable" && name != null && (restrictedNames.includes(name.toLowerCase()) || smell.hasDebugInName(name.toLowerCase())) && value == true){
                smell.triggerAlarm (fileName, MSG, lineno, WARNING_MSG);
            }
            else if(tokenType == 'dict' && token.hasOwnProperty('pairs')){
                for (const pair of token.pairs){
                    if(restrictedNames.includes(pair[0]) && pair[1] == true){
                        smell.triggerAlarm (fileName, MSG, lineno, WARNING_MSG);
                    }
                }
            }
            else if(tokenType == "function_call" && token.hasOwnProperty("keywords")) {
                for(const keyword of token.keywords){
                    if(restrictedNames.includes(keyword[0]) && keyword[1] == true){
                        smell.triggerAlarm (fileName, MSG, lineno, WARNING_MSG);
                    }
                }
            }
        } catch(error){
            console.log(error);
        }      
    },
    hasDebugInName: (variableName) => {
        const restrictedNames = ['debug','debug_propagate_exceptions','propagate_exceptions','PROPAGATE_EXCEPTIONS'];
        for (const name of restrictedNames){
            if (variableName.search(name)) return true 
        }

        return false
    },
    
    triggerAlarm: (fileName, MSG, lineno, WARNING_MSG) => {
        let backslashSplittedFilePathLength = fileName.split("/").length
        let filenameFromPath = fileName.split("/")[backslashSplittedFilePathLength - 1]
        
        vscode.window.showWarningMessage(MSG +" : "+ filenameFromPath+":"+lineno);
        console.log( "\u001b[1;31m"+"warning: "+MSG +"  location:"+ fileName+":"+lineno);
        fs.appendFileSync('smell-spotter/warning-logs/project_warnings.csv', fileName+" ,"+WARNING_MSG+"\n");
    }
}

module.exports = smell;

File: py-scripts/filters/debugflag.py
class DebugFlag:
    '''This is the class for detecting debug flag set to true in code'''

    def __init__(self):
        self.restricted_names = ['debug','debug_propagate_exceptions','propagate_exceptions']
        self.warning_message = 'deployment with debug flag set to true'

    def detect_smell(self, token, src_file):
        try:
            if token.__contains__("line"): lineno = token["line"]
            if token.__contains__("type"): tokenType = token["type"]
            if token.__contains__("name"): name = token["name"]
            if token.__contains__("value"): value = token["value"]
            
            
            if tokenType == "variable" and name is not None and (name.lower() in self.restricted_names or self.has_debug_in_name(name.lower())) and value is True: 
                self.trigger_alarm(src_file, lineno)


            elif tokenType == "function_call" and token.__contains__("keywords"):
                
                for keyword in token["keywords"]:
                    if(keyword[0] is not None and isinstance(keyword[0], str) and keyword[0].lower() in self.restricted_names and keyword[1] is True): 
                        self.trigger_alarm(src_file, lineno)


            elif tokenType == "dict" and token.__contains__("pairs"): 
                
                for pair in token['pairs']:
                    if pair[0] in self.restricted_names and pair[1] is True: 
                        self.trigger_alarm(src_file, lineno)
        
        except Exception as error: 
            print("debug error:"+str(error))

    def has_debug_in_name(self, var_name):
        for name in self.restricted_names:
            if name in var_name.lower().strip(): 
                return True
        
        return False


    def trigger_alarm(self, src_file, lineno):
        print(src_file +":"+ str(lineno)+" ,"+self.warning_message)
    